### PR TITLE
Allow multiple auth providers with the same URL, adjust UI

### PR DESCRIPTION
**Pull request by pjlast**
This PR removes the restrictions of having one Auth provider per URL, and instead allows one auth provider for each URL+ClientID combination.

The URL restriction came from the way our old permissions table worked, but with the `user_repo_permissions` table, this restriction no longer applies as permissions are stored against individual external accounts.

This has also become an increasingly common ask from customers, and it also eases the transitioning when switching from one auth provider to another, like switching from traditional GitHub OAuth to GitHub App Oauth.

One caveat that remains: Having multiple auth providers for the same URL does not play well with repo-centric permissions syncing, as we cannot from the repo token's perspective determine which user external account a permission should be bound to. There isn't a clean workaround to make this work, and the suggestion is to turn off repo-centric permissions syncing when using multiple auth providers to the same URL.

The idea is to have a follow-up PR that allows us to disable repo-centric permissions syncs for individual code host connections.

Having two auth providers and account connections that both point to github.com:

image

Some follow-up tasks:

- [ ] Change [this](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/repo-updater/internal/authz/perms_syncer.go?L432%3A2-438%3A40) code so that auth providers are grouped by UniqueID and not ServiceID
  - This will require changes to each individual auth provider. But it would allow us to not sync permissions using an external account that no longer has an active auth provider.
- [ ] Allow users to delete external accounts that are no longer associated with a valid auth provider
- [ ] Add a code host connection option that allows disabling repo-permissions syncing on a code host connection level

## Test plan

Lots of tests updated and adjusted to fit the new criteria.

### CODE REVIEW

File: enterprise/cmd/frontend/internal/auth/azureoauth/provider.go [hunk 1]
```diff
@@ -84,10 +85,14 @@ func parseConfig(logger log.Logger, cfg conftypes.SiteConfigQuerier, db database
...
 		}
 
-		// Currently Azure Dev Ops will work only against https://dev.azure.com. If we have more
-		// than one configuration for Azure Dev Ops auth provider, we want to fail early.
+		var configured bool
+		for _, clientID := range configuredIDs {
+			if clientID == pr.AzureDevOps.ClientID {
+				configured = true
+			}
+		}
```
Comment (kopancek) [inline enterprise/cmd/frontend/internal/auth/azureoauth/provider.go:93]: As much as I dislike this code and wish there was a builtin `Contains` or similar on the slice, we should probably break out of the loop when we find a match.

```suggestion
		for _, clientID := range configuredIDs {
			if clientID == pr.AzureDevOps.ClientID {
				configured = true
				break
			}
		}
```

Unless we use a `Set` as suggested above.

File: enterprise/cmd/frontend/internal/auth/azureoauth/provider.go [hunk 2]
```diff
@@ -84,10 +85,14 @@ func parseConfig(logger log.Logger, cfg conftypes.SiteConfigQuerier, db database
...
 		if configured {
-			problems = append(problems, conf.NewSiteProblem("Cannot have more than one auth provider for Azure Dev Ops, only the first one will be used"))
+			problems = append(problems, conf.NewSiteProblem(fmt.Sprintf("Cannot have more than one auth provider for Azure Dev Ops with Client ID %q, only the first one will be used", pr.AzureDevOps.ClientID)))
```
Comment (kopancek) [inline enterprise/cmd/frontend/internal/auth/azureoauth/provider.go:95]:
```suggestion
			problems = append(problems, conf.NewSiteProblem(fmt.Sprintf("Cannot have more than one auth provider for Azure Dev Ops with the same Client ID %q, only the first one will be used", pr.AzureDevOps.ClientID)))
```

File: enterprise/cmd/frontend/internal/auth/azureoauth/provider.go [hunk 3]
```diff
@@ -69,7 +69,8 @@ func parseConfig(logger log.Logger, cfg conftypes.SiteConfigQuerier, db database
 		return ps, problems
 	}
 
-	var configured bool
+	var configuredIDs []string
```
Comment (kopancek) [inline enterprise/cmd/frontend/internal/auth/azureoauth/provider.go:72]: If we use a `Set<string>` instead, we can check easily if the set contains an element instead of doing a for loop.

File: client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx [hunk 4]
```diff
@@ -41,18 +41,26 @@ const getNormalizedAccount = (
...
         name,
     }
 
-    if (account?.publicAccountData) {
-        normalizedAccount.external = {
-            id: account.id,
-            ...account.publicAccountData,
+    for (let i = 0; i < providerAccounts.length; i++) {
+        const pAcc = providerAccounts[i]
+        if (pAcc && pAcc.clientID === authProvider.clientID) {
```
Comment (sashaostrikov) [inline client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:57]: nit: Not a JS expert here, but I think we can simplify this a bit, using [`find`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find).

File: client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx [hunk 5]
```diff
@@ -76,7 +84,7 @@ export const ExternalAccountsSignIn: React.FunctionComponent<React.PropsWithChil
                     if (normAccount) {
                         return (
                             <li
-                                key={authProvider.serviceID}
+                                key={normAccount.external ? normAccount.external.id : authProvider.serviceID}
```
Comment (sashaostrikov) [inline client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:87]: Is this change necessary? Can we not have ID collisions between `external.id` and `serviceID`? 🤔

File: client/web/src/user/settings/auth/UserSettingsSecurityPage.tsx [hunk 6]
```diff
@@ -92,8 +92,11 @@ export const UserSettingsSecurityPage: React.FunctionComponent<React.PropsWithCh
     }
 
     // auth providers by service ID
-    const accountByServiceID = accounts.fetched?.reduce((accumulator: AccountByServiceID, account) => {
-        accumulator[account.serviceID] = account
+    const accountsByServiceID = accounts.fetched?.reduce((accumulator: AccountsByServiceID, account) => {
+        let accountMap = accumulator[account.serviceID] ?? []
+        accountMap.push(account)
+        accumulator[account.serviceID] = accountMap
```
Comment (sashaostrikov) [inline client/web/src/user/settings/auth/UserSettingsSecurityPage.tsx:98]: Nice try, Mr. Array.

```suggestion
        let accountArray = accumulator[account.serviceID] ?? []
        accountArray.push(account)
        accumulator[account.serviceID] = accountArray
```

File: enterprise/cmd/frontend/internal/auth/bitbucketcloudoauth/config.go [hunk 7]
```diff
@@ -61,16 +63,16 @@ func parseConfig(logger log.Logger, cfg conftypes.SiteConfigQuerier, db database
 			continue
 		}
 
-		if _, ok := configured[provider.ServiceID]; ok {
-			problems = append(problems, conf.NewSiteProblems(fmt.Sprintf(`Cannot have more than one auth provider with url %q, only the first one will be used`, provider.ServiceID))...)
+		if existingProviders.Has(provider.CachedInfo().UniqueID()) {
+			problems = append(problems, conf.NewSiteProblems(fmt.Sprintf(`Cannot have more than one auth provider with url %q and client ID %q, only the first one will be used`, provider.ServiceID, provider.CachedInfo().ClientID))...)
```
Comment (sashaostrikov) [inline enterprise/cmd/frontend/internal/auth/bitbucketcloudoauth/config.go:67]: nit: let's also specify code host type in the message (same as with ADO)
```suggestion
			problems = append(problems, conf.NewSiteProblems(fmt.Sprintf(`Cannot have more than one auth provider for Bitbucket Cloud with url %q and client ID %q, only the first one will be used`, provider.ServiceID, provider.CachedInfo().ClientID))...)
```

File: enterprise/cmd/frontend/internal/auth/gerrit/config.go [hunk 8]
```diff
@@ -35,19 +36,20 @@ type Provider struct {
...
 			continue
 		}
 
 		provider := parseProvider(pr.Gerrit)
-		if _, ok := seen[provider.ServiceID]; !ok {
-			ps = append(ps, provider)
-			seen[provider.ServiceID] = struct{}{}
-		} else {
+		if existingProviders.Has(provider.CachedInfo().UniqueID()) {
 			problems = append(problems, conf.NewSiteProblem(fmt.Sprintf("Cannot have more than one auth provider with url %q", provider.ServiceID)))
```
Comment (sashaostrikov) [inline enterprise/cmd/frontend/internal/auth/gerrit/config.go:47]:
```suggestion
			problems = append(problems, conf.NewSiteProblem(fmt.Sprintf("Cannot have more than one auth provider for Gerrit with url %q", provider.ServiceID)))
```

File: enterprise/cmd/frontend/internal/auth/githuboauth/config.go [hunk 9]
```diff
@@ -51,29 +52,29 @@ type Provider struct {
...
-			ps = append(ps, Provider{
-				GitHubAuthProvider: pr.Github,
-				Provider:           provider,
-			})
+		if provider == nil {
+			continue
 		}
+
+		if existingProviders.Has(provider.CachedInfo().UniqueID()) {
+			problems = append(problems, conf.NewSiteProblems(fmt.Sprintf(`Cannot have more than one auth provider with url %q and client ID %q, only the first one will be used`, provider.ServiceID, provider.CachedInfo().ClientID))...)
```
Comment (sashaostrikov) [inline enterprise/cmd/frontend/internal/auth/githuboauth/config.go:68]:
```suggestion
			problems = append(problems, conf.NewSiteProblems(fmt.Sprintf(`Cannot have more than one auth provider for GitHub with url %q and client ID %q, only the first one will be used`, provider.ServiceID, provider.CachedInfo().ClientID))...)
```

File: enterprise/cmd/frontend/internal/auth/gitlaboauth/config.go [hunk 10]
```diff
@@ -70,24 +72,22 @@ func parseConfig(logger log.Logger, cfg conftypes.SiteConfigQuerier, db database
...
-			}
-			if alreadyExists {
-				continue
-			}
+		if provider == nil {
+			continue
 		}
+
+		if existingProviders.Has(provider.CachedInfo().UniqueID()) {
+			problems = append(problems, conf.NewSiteProblems(fmt.Sprintf(`Cannot have more than one auth provider with url %q and client ID %q, only the first one will be used`, provider.ServiceID, provider.CachedInfo().ClientID))...)
```
Comment (sashaostrikov) [inline enterprise/cmd/frontend/internal/auth/gitlaboauth/config.go:81]:
```suggestion
			problems = append(problems, conf.NewSiteProblems(fmt.Sprintf(`Cannot have more than one auth provider for GitLab with url %q and client ID %q, only the first one will be used`, provider.ServiceID, provider.CachedInfo().ClientID))...)
```

**Review by sashaostrikov — APPROVED** [review]
LGTM! Left some comments, none of them are critical

File: client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx (hunk 4 above)
Comment (pjlast) [reply client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:57]: Ooo nice! Today I learned.

File: client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx (hunk 5 above)
Comment (pjlast) [reply client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:87]: We can have ID collisions on `serviceID` , since `serviceID` is just the URL of the auth provider, and if we have multiple auth providers for the same URL, they will clash.

So we instead use `external.id`, and in case it doesn't exist for some reason, like and account not having external data, we use the `serviceID` (since I expect these kinds of accounts to not allow multiple providers per URL anyways)

File: client/web/src/user/settings/auth/UserSettingsSecurityPage.tsx (hunk 6 above)
Comment (pjlast) [reply client/web/src/user/settings/auth/UserSettingsSecurityPage.tsx:98]: And I would have gotten away with it to, if it weren't for that meddling kid! - Mr. Array, probably

File: client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx (hunk 5 above)
Comment (sashaostrikov) [reply client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:87]: nice, thanks for the explanation and sorry for a dumb question 😁
